// ddex-api/src/models/trade.ts
import { BigNumber } from "bignumber.js";
import { TradeJson } from "../types"

export class Trade {
  id: string
  maker: string
  taker: string
  amount: BigNumber
  price: BigNumber
  takerPrice: BigNumber
  feeAmount: BigNumber
  buyer: string
  seller: string
  pair: string
  status: string
  executedAt: Date
  createdAt: Date
  transactionId: string

  static fromJSON(tradeJson: TradeJson): Trade {
    const trade = new Trade()
    trade.id = tradeJson.id
    trade.maker = tradeJson.maker
    trade.taker = tradeJson.taker
    trade.price = new BigNumber(tradeJson.price)
    trade.amount = new BigNumber(tradeJson.amount)
    trade.takerPrice = new BigNumber(tradeJson.takerPrice)
    trade.feeAmount = new BigNumber(tradeJson.feeAmount)
    trade.buyer = tradeJson.buyer
    trade.seller = tradeJson.seller
    trade.pair = tradeJson.pair
    trade.status = tradeJson.status
    trade.executedAt = new Date(tradeJson.executedAt)
    trade.createdAt = new Date(tradeJson.createdAt)
    trade.transactionId = tradeJson.transactionId
    return trade
  }
}
