// 0x.js/lib/src/contract_wrappers/generated/exchange.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * This file is auto-generated using abi-gen. Don't edit directly.
 * Templates can be found at https://github.com/0xProject/0x-monorepo/tree/development/packages/contract_templates.
 */
// tslint:disable:no-consecutive-blank-lines
// tslint:disable-next-line:no-unused-variable
var base_contract_1 = require("@0xproject/base-contract");
var utils_1 = require("@0xproject/utils");
var ethersContracts = require("ethers-contracts");
var _ = require("lodash");
var ExchangeEvents;
(function (ExchangeEvents) {
    ExchangeEvents["LogFill"] = "LogFill";
    ExchangeEvents["LogCancel"] = "LogCancel";
    ExchangeEvents["LogError"] = "LogError";
})(ExchangeEvents = exports.ExchangeEvents || (exports.ExchangeEvents = {}));
// tslint:disable:no-parameter-reassignment
var ExchangeContract = /** @class */ (function (_super) {
    __extends(ExchangeContract, _super);
    function ExchangeContract(web3Wrapper, abi, address) {
        var _this = _super.call(this, web3Wrapper, abi, address) || this;
        _this.isRoundingError = {
            callAsync: function (numerator, denominator, target, txData, defaultBlock) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, callData, rawCallResult, outputAbi, outputParamsTypes, resultArray, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'isRoundingError' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [numerator, denominator, target], base_contract_1.BaseContract._bigNumberToString.bind(this)), numerator = _a[0], denominator = _a[1], target = _a[2];
                                encodedData = self._ethersInterface.functions.isRoundingError(numerator, denominator, target).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync({
                                        data: encodedData,
                                    })];
                            case 1:
                                callData = _b.sent();
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callData, defaultBlock)];
                            case 2:
                                rawCallResult = _b.sent();
                                outputAbi = _.find(this.abi, { name: 'isRoundingError' }).outputs;
                                outputParamsTypes = _.map(outputAbi, 'type');
                                resultArray = ethersContracts.Interface.decodeParams(outputParamsTypes, rawCallResult);
                                resultArray = base_contract_1.BaseContract._transformABIData(outputAbi, resultArray, base_contract_1.BaseContract._lowercaseAddress.bind(this));
                                return [2 /*return*/, resultArray[0]];
                        }
                    });
                });
            },
        };
        _this.filled = {
            callAsync: function (index_0, txData, defaultBlock) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, callData, rawCallResult, outputAbi, outputParamsTypes, resultArray;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'filled' }).inputs;
                                index_0 = base_contract_1.BaseContract._transformABIData(inputAbi, [index_0], base_contract_1.BaseContract._bigNumberToString.bind(this))[0];
                                encodedData = self._ethersInterface.functions.filled(index_0).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync({
                                        data: encodedData,
                                    })];
                            case 1:
                                callData = _a.sent();
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callData, defaultBlock)];
                            case 2:
                                rawCallResult = _a.sent();
                                outputAbi = _.find(this.abi, { name: 'filled' }).outputs;
                                outputParamsTypes = _.map(outputAbi, 'type');
                                resultArray = ethersContracts.Interface.decodeParams(outputParamsTypes, rawCallResult);
                                resultArray = base_contract_1.BaseContract._transformABIData(outputAbi, resultArray, base_contract_1.BaseContract._lowercaseAddress.bind(this));
                                return [2 /*return*/, resultArray[0]];
                        }
                    });
                });
            },
        };
        _this.cancelled = {
            callAsync: function (index_0, txData, defaultBlock) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, callData, rawCallResult, outputAbi, outputParamsTypes, resultArray;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'cancelled' }).inputs;
                                index_0 = base_contract_1.BaseContract._transformABIData(inputAbi, [index_0], base_contract_1.BaseContract._bigNumberToString.bind(this))[0];
                                encodedData = self._ethersInterface.functions.cancelled(index_0).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync({
                                        data: encodedData,
                                    })];
                            case 1:
                                callData = _a.sent();
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callData, defaultBlock)];
                            case 2:
                                rawCallResult = _a.sent();
                                outputAbi = _.find(this.abi, { name: 'cancelled' }).outputs;
                                outputParamsTypes = _.map(outputAbi, 'type');
                                resultArray = ethersContracts.Interface.decodeParams(outputParamsTypes, rawCallResult);
                                resultArray = base_contract_1.BaseContract._transformABIData(outputAbi, resultArray, base_contract_1.BaseContract._lowercaseAddress.bind(this));
                                return [2 /*return*/, resultArray[0]];
                        }
                    });
                });
            },
        };
        _this.fillOrdersUpTo = {
            sendTransactionAsync: function (orderAddresses, orderValues, fillTakerTokenAmount, shouldThrowOnInsufficientBalanceOrAllowance, v, r, s, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, txDataWithDefaults, txHash, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'fillOrdersUpTo' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [
                                    orderAddresses,
                                    orderValues,
                                    fillTakerTokenAmount,
                                    shouldThrowOnInsufficientBalanceOrAllowance,
                                    v,
                                    r,
                                    s,
                                ], base_contract_1.BaseContract._bigNumberToString.bind(this)), orderAddresses = _a[0], orderValues = _a[1], fillTakerTokenAmount = _a[2], shouldThrowOnInsufficientBalanceOrAllowance = _a[3], v = _a[4], r = _a[5], s = _a[6];
                                encodedData = this._ethersInterface.functions.fillOrdersUpTo(orderAddresses, orderValues, fillTakerTokenAmount, shouldThrowOnInsufficientBalanceOrAllowance, v, r, s).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync(__assign({}, txData, { data: encodedData }), self.fillOrdersUpTo.estimateGasAsync.bind(self, orderAddresses, orderValues, fillTakerTokenAmount, shouldThrowOnInsufficientBalanceOrAllowance, v, r, s))];
                            case 1:
                                txDataWithDefaults = _b.sent();
                                return [4 /*yield*/, this._web3Wrapper.sendTransactionAsync(txDataWithDefaults)];
                            case 2:
                                txHash = _b.sent();
                                return [2 /*return*/, txHash];
                        }
                    });
                });
            },
            estimateGasAsync: function (orderAddresses, orderValues, fillTakerTokenAmount, shouldThrowOnInsufficientBalanceOrAllowance, v, r, s, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, txDataWithDefaults, gas, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'fillOrdersUpTo' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [
                                    orderAddresses,
                                    orderValues,
                                    fillTakerTokenAmount,
                                    shouldThrowOnInsufficientBalanceOrAllowance,
                                    v,
                                    r,
                                    s,
                                ], base_contract_1.BaseContract._bigNumberToString.bind(this)), orderAddresses = _a[0], orderValues = _a[1], fillTakerTokenAmount = _a[2], shouldThrowOnInsufficientBalanceOrAllowance = _a[3], v = _a[4], r = _a[5], s = _a[6];
                                encodedData = this._ethersInterface.functions.fillOrdersUpTo(orderAddresses, orderValues, fillTakerTokenAmount, shouldThrowOnInsufficientBalanceOrAllowance, v, r, s).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync(__assign({}, txData, { data: encodedData }))];
                            case 1:
                                txDataWithDefaults = _b.sent();
                                return [4 /*yield*/, this._web3Wrapper.estimateGasAsync(txDataWithDefaults)];
                            case 2:
                                gas = _b.sent();
                                return [2 /*return*/, gas];
                        }
                    });
                });
            },
            getABIEncodedTransactionData: function (orderAddresses, orderValues, fillTakerTokenAmount, shouldThrowOnInsufficientBalanceOrAllowance, v, r, s, txData) {
                if (txData === void 0) { txData = {}; }
                var self = this;
                var inputAbi = _.find(this.abi, { name: 'fillOrdersUpTo' }).inputs;
                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [
                    orderAddresses,
                    orderValues,
                    fillTakerTokenAmount,
                    shouldThrowOnInsufficientBalanceOrAllowance,
                    v,
                    r,
                    s,
                ], base_contract_1.BaseContract._bigNumberToString.bind(this)), orderAddresses = _a[0], orderValues = _a[1], fillTakerTokenAmount = _a[2], shouldThrowOnInsufficientBalanceOrAllowance = _a[3], v = _a[4], r = _a[5], s = _a[6];
                var abiEncodedTransactionData = this._ethersInterface.functions.fillOrdersUpTo(orderAddresses, orderValues, fillTakerTokenAmount, shouldThrowOnInsufficientBalanceOrAllowance, v, r, s).data;
                return abiEncodedTransactionData;
                var _a;
            },
            callAsync: function (orderAddresses, orderValues, fillTakerTokenAmount, shouldThrowOnInsufficientBalanceOrAllowance, v, r, s, txData, defaultBlock) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, callData, rawCallResult, outputAbi, outputParamsTypes, resultArray, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'fillOrdersUpTo' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [
                                    orderAddresses,
                                    orderValues,
                                    fillTakerTokenAmount,
                                    shouldThrowOnInsufficientBalanceOrAllowance,
                                    v,
                                    r,
                                    s,
                                ], base_contract_1.BaseContract._bigNumberToString.bind(this)), orderAddresses = _a[0], orderValues = _a[1], fillTakerTokenAmount = _a[2], shouldThrowOnInsufficientBalanceOrAllowance = _a[3], v = _a[4], r = _a[5], s = _a[6];
                                encodedData = self._ethersInterface.functions.fillOrdersUpTo(orderAddresses, orderValues, fillTakerTokenAmount, shouldThrowOnInsufficientBalanceOrAllowance, v, r, s).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync({
                                        data: encodedData,
                                    })];
                            case 1:
                                callData = _b.sent();
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callData, defaultBlock)];
                            case 2:
                                rawCallResult = _b.sent();
                                outputAbi = _.find(this.abi, { name: 'fillOrdersUpTo' }).outputs;
                                outputParamsTypes = _.map(outputAbi, 'type');
                                resultArray = ethersContracts.Interface.decodeParams(outputParamsTypes, rawCallResult);
                                resultArray = base_contract_1.BaseContract._transformABIData(outputAbi, resultArray, base_contract_1.BaseContract._lowercaseAddress.bind(this));
                                return [2 /*return*/, resultArray[0]];
                        }
                    });
                });
            },
        };
        _this.cancelOrder = {
            sendTransactionAsync: function (orderAddresses, orderValues, cancelTakerTokenAmount, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, txDataWithDefaults, txHash, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'cancelOrder' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [orderAddresses, orderValues, cancelTakerTokenAmount], base_contract_1.BaseContract._bigNumberToString.bind(this)), orderAddresses = _a[0], orderValues = _a[1], cancelTakerTokenAmount = _a[2];
                                encodedData = this._ethersInterface.functions.cancelOrder(orderAddresses, orderValues, cancelTakerTokenAmount).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync(__assign({}, txData, { data: encodedData }), self.cancelOrder.estimateGasAsync.bind(self, orderAddresses, orderValues, cancelTakerTokenAmount))];
                            case 1:
                                txDataWithDefaults = _b.sent();
                                return [4 /*yield*/, this._web3Wrapper.sendTransactionAsync(txDataWithDefaults)];
                            case 2:
                                txHash = _b.sent();
                                return [2 /*return*/, txHash];
                        }
                    });
                });
            },
            estimateGasAsync: function (orderAddresses, orderValues, cancelTakerTokenAmount, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, txDataWithDefaults, gas, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'cancelOrder' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [orderAddresses, orderValues, cancelTakerTokenAmount], base_contract_1.BaseContract._bigNumberToString.bind(this)), orderAddresses = _a[0], orderValues = _a[1], cancelTakerTokenAmount = _a[2];
                                encodedData = this._ethersInterface.functions.cancelOrder(orderAddresses, orderValues, cancelTakerTokenAmount).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync(__assign({}, txData, { data: encodedData }))];
                            case 1:
                                txDataWithDefaults = _b.sent();
                                return [4 /*yield*/, this._web3Wrapper.estimateGasAsync(txDataWithDefaults)];
                            case 2:
                                gas = _b.sent();
                                return [2 /*return*/, gas];
                        }
                    });
                });
            },
            getABIEncodedTransactionData: function (orderAddresses, orderValues, cancelTakerTokenAmount, txData) {
                if (txData === void 0) { txData = {}; }
                var self = this;
                var inputAbi = _.find(this.abi, { name: 'cancelOrder' }).inputs;
                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [orderAddresses, orderValues, cancelTakerTokenAmount], base_contract_1.BaseContract._bigNumberToString.bind(this)), orderAddresses = _a[0], orderValues = _a[1], cancelTakerTokenAmount = _a[2];
                var abiEncodedTransactionData = this._ethersInterface.functions.cancelOrder(orderAddresses, orderValues, cancelTakerTokenAmount).data;
                return abiEncodedTransactionData;
                var _a;
            },
            callAsync: function (orderAddresses, orderValues, cancelTakerTokenAmount, txData, defaultBlock) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, callData, rawCallResult, outputAbi, outputParamsTypes, resultArray, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'cancelOrder' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [orderAddresses, orderValues, cancelTakerTokenAmount], base_contract_1.BaseContract._bigNumberToString.bind(this)), orderAddresses = _a[0], orderValues = _a[1], cancelTakerTokenAmount = _a[2];
                                encodedData = self._ethersInterface.functions.cancelOrder(orderAddresses, orderValues, cancelTakerTokenAmount).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync({
                                        data: encodedData,
                                    })];
                            case 1:
                                callData = _b.sent();
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callData, defaultBlock)];
                            case 2:
                                rawCallResult = _b.sent();
                                outputAbi = _.find(this.abi, { name: 'cancelOrder' }).outputs;
                                outputParamsTypes = _.map(outputAbi, 'type');
                                resultArray = ethersContracts.Interface.decodeParams(outputParamsTypes, rawCallResult);
                                resultArray = base_contract_1.BaseContract._transformABIData(outputAbi, resultArray, base_contract_1.BaseContract._lowercaseAddress.bind(this));
                                return [2 /*return*/, resultArray[0]];
                        }
                    });
                });
            },
        };
        _this.ZRX_TOKEN_CONTRACT = {
            callAsync: function (txData, defaultBlock) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, callData, rawCallResult, outputAbi, outputParamsTypes, resultArray;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'ZRX_TOKEN_CONTRACT' }).inputs;
                                base_contract_1.BaseContract._transformABIData(inputAbi, [], base_contract_1.BaseContract._bigNumberToString.bind(this));
                                encodedData = self._ethersInterface.functions.ZRX_TOKEN_CONTRACT().data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync({
                                        data: encodedData,
                                    })];
                            case 1:
                                callData = _a.sent();
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callData, defaultBlock)];
                            case 2:
                                rawCallResult = _a.sent();
                                outputAbi = _.find(this.abi, { name: 'ZRX_TOKEN_CONTRACT' }).outputs;
                                outputParamsTypes = _.map(outputAbi, 'type');
                                resultArray = ethersContracts.Interface.decodeParams(outputParamsTypes, rawCallResult);
                                resultArray = base_contract_1.BaseContract._transformABIData(outputAbi, resultArray, base_contract_1.BaseContract._lowercaseAddress.bind(this));
                                return [2 /*return*/, resultArray[0]];
                        }
                    });
                });
            },
        };
        _this.batchFillOrKillOrders = {
            sendTransactionAsync: function (orderAddresses, orderValues, fillTakerTokenAmounts, v, r, s, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, txDataWithDefaults, txHash, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'batchFillOrKillOrders' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [orderAddresses, orderValues, fillTakerTokenAmounts, v, r, s], base_contract_1.BaseContract._bigNumberToString.bind(this)), orderAddresses = _a[0], orderValues = _a[1], fillTakerTokenAmounts = _a[2], v = _a[3], r = _a[4], s = _a[5];
                                encodedData = this._ethersInterface.functions.batchFillOrKillOrders(orderAddresses, orderValues, fillTakerTokenAmounts, v, r, s).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync(__assign({}, txData, { data: encodedData }), self.batchFillOrKillOrders.estimateGasAsync.bind(self, orderAddresses, orderValues, fillTakerTokenAmounts, v, r, s))];
                            case 1:
                                txDataWithDefaults = _b.sent();
                                return [4 /*yield*/, this._web3Wrapper.sendTransactionAsync(txDataWithDefaults)];
                            case 2:
                                txHash = _b.sent();
                                return [2 /*return*/, txHash];
                        }
                    });
                });
            },
            estimateGasAsync: function (orderAddresses, orderValues, fillTakerTokenAmounts, v, r, s, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, txDataWithDefaults, gas, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'batchFillOrKillOrders' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [orderAddresses, orderValues, fillTakerTokenAmounts, v, r, s], base_contract_1.BaseContract._bigNumberToString.bind(this)), orderAddresses = _a[0], orderValues = _a[1], fillTakerTokenAmounts = _a[2], v = _a[3], r = _a[4], s = _a[5];
                                encodedData = this._ethersInterface.functions.batchFillOrKillOrders(orderAddresses, orderValues, fillTakerTokenAmounts, v, r, s).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync(__assign({}, txData, { data: encodedData }))];
                            case 1:
                                txDataWithDefaults = _b.sent();
                                return [4 /*yield*/, this._web3Wrapper.estimateGasAsync(txDataWithDefaults)];
                            case 2:
                                gas = _b.sent();
                                return [2 /*return*/, gas];
                        }
                    });
                });
            },
            getABIEncodedTransactionData: function (orderAddresses, orderValues, fillTakerTokenAmounts, v, r, s, txData) {
                if (txData === void 0) { txData = {}; }
                var self = this;
                var inputAbi = _.find(this.abi, { name: 'batchFillOrKillOrders' }).inputs;
                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [orderAddresses, orderValues, fillTakerTokenAmounts, v, r, s], base_contract_1.BaseContract._bigNumberToString.bind(this)), orderAddresses = _a[0], orderValues = _a[1], fillTakerTokenAmounts = _a[2], v = _a[3], r = _a[4], s = _a[5];
                var abiEncodedTransactionData = this._ethersInterface.functions.batchFillOrKillOrders(orderAddresses, orderValues, fillTakerTokenAmounts, v, r, s).data;
                return abiEncodedTransactionData;
                var _a;
            },
        };
        _this.fillOrKillOrder = {
            sendTransactionAsync: function (orderAddresses, orderValues, fillTakerTokenAmount, v, r, s, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, txDataWithDefaults, txHash, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'fillOrKillOrder' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [orderAddresses, orderValues, fillTakerTokenAmount, v, r, s], base_contract_1.BaseContract._bigNumberToString.bind(this)), orderAddresses = _a[0], orderValues = _a[1], fillTakerTokenAmount = _a[2], v = _a[3], r = _a[4], s = _a[5];
                                encodedData = this._ethersInterface.functions.fillOrKillOrder(orderAddresses, orderValues, fillTakerTokenAmount, v, r, s).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync(__assign({}, txData, { data: encodedData }), self.fillOrKillOrder.estimateGasAsync.bind(self, orderAddresses, orderValues, fillTakerTokenAmount, v, r, s))];
                            case 1:
                                txDataWithDefaults = _b.sent();
                                return [4 /*yield*/, this._web3Wrapper.sendTransactionAsync(txDataWithDefaults)];
                            case 2:
                                txHash = _b.sent();
                                return [2 /*return*/, txHash];
                        }
                    });
                });
            },
            estimateGasAsync: function (orderAddresses, orderValues, fillTakerTokenAmount, v, r, s, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, txDataWithDefaults, gas, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'fillOrKillOrder' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [orderAddresses, orderValues, fillTakerTokenAmount, v, r, s], base_contract_1.BaseContract._bigNumberToString.bind(this)), orderAddresses = _a[0], orderValues = _a[1], fillTakerTokenAmount = _a[2], v = _a[3], r = _a[4], s = _a[5];
                                encodedData = this._ethersInterface.functions.fillOrKillOrder(orderAddresses, orderValues, fillTakerTokenAmount, v, r, s).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync(__assign({}, txData, { data: encodedData }))];
                            case 1:
                                txDataWithDefaults = _b.sent();
                                return [4 /*yield*/, this._web3Wrapper.estimateGasAsync(txDataWithDefaults)];
                            case 2:
                                gas = _b.sent();
                                return [2 /*return*/, gas];
                        }
                    });
                });
            },
            getABIEncodedTransactionData: function (orderAddresses, orderValues, fillTakerTokenAmount, v, r, s, txData) {
                if (txData === void 0) { txData = {}; }
                var self = this;
                var inputAbi = _.find(this.abi, { name: 'fillOrKillOrder' }).inputs;
                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [orderAddresses, orderValues, fillTakerTokenAmount, v, r, s], base_contract_1.BaseContract._bigNumberToString.bind(this)), orderAddresses = _a[0], orderValues = _a[1], fillTakerTokenAmount = _a[2], v = _a[3], r = _a[4], s = _a[5];
                var abiEncodedTransactionData = this._ethersInterface.functions.fillOrKillOrder(orderAddresses, orderValues, fillTakerTokenAmount, v, r, s).data;
                return abiEncodedTransactionData;
                var _a;
            },
        };
        _this.getUnavailableTakerTokenAmount = {
            callAsync: function (orderHash, txData, defaultBlock) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, callData, rawCallResult, outputAbi, outputParamsTypes, resultArray;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'getUnavailableTakerTokenAmount' }).inputs;
                                orderHash = base_contract_1.BaseContract._transformABIData(inputAbi, [orderHash], base_contract_1.BaseContract._bigNumberToString.bind(this))[0];
                                encodedData = self._ethersInterface.functions.getUnavailableTakerTokenAmount(orderHash).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync({
                                        data: encodedData,
                                    })];
                            case 1:
                                callData = _a.sent();
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callData, defaultBlock)];
                            case 2:
                                rawCallResult = _a.sent();
                                outputAbi = _.find(this.abi, { name: 'getUnavailableTakerTokenAmount' }).outputs;
                                outputParamsTypes = _.map(outputAbi, 'type');
                                resultArray = ethersContracts.Interface.decodeParams(outputParamsTypes, rawCallResult);
                                resultArray = base_contract_1.BaseContract._transformABIData(outputAbi, resultArray, base_contract_1.BaseContract._lowercaseAddress.bind(this));
                                return [2 /*return*/, resultArray[0]];
                        }
                    });
                });
            },
        };
        _this.isValidSignature = {
            callAsync: function (signer, hash, v, r, s, txData, defaultBlock) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, callData, rawCallResult, outputAbi, outputParamsTypes, resultArray, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'isValidSignature' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [signer, hash, v, r, s], base_contract_1.BaseContract._bigNumberToString.bind(this)), signer = _a[0], hash = _a[1], v = _a[2], r = _a[3], s = _a[4];
                                encodedData = self._ethersInterface.functions.isValidSignature(signer, hash, v, r, s).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync({
                                        data: encodedData,
                                    })];
                            case 1:
                                callData = _b.sent();
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callData, defaultBlock)];
                            case 2:
                                rawCallResult = _b.sent();
                                outputAbi = _.find(this.abi, { name: 'isValidSignature' }).outputs;
                                outputParamsTypes = _.map(outputAbi, 'type');
                                resultArray = ethersContracts.Interface.decodeParams(outputParamsTypes, rawCallResult);
                                resultArray = base_contract_1.BaseContract._transformABIData(outputAbi, resultArray, base_contract_1.BaseContract._lowercaseAddress.bind(this));
                                return [2 /*return*/, resultArray[0]];
                        }
                    });
                });
            },
        };
        _this.getPartialAmount = {
            callAsync: function (numerator, denominator, target, txData, defaultBlock) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, callData, rawCallResult, outputAbi, outputParamsTypes, resultArray, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'getPartialAmount' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [numerator, denominator, target], base_contract_1.BaseContract._bigNumberToString.bind(this)), numerator = _a[0], denominator = _a[1], target = _a[2];
                                encodedData = self._ethersInterface.functions.getPartialAmount(numerator, denominator, target).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync({
                                        data: encodedData,
                                    })];
                            case 1:
                                callData = _b.sent();
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callData, defaultBlock)];
                            case 2:
                                rawCallResult = _b.sent();
                                outputAbi = _.find(this.abi, { name: 'getPartialAmount' }).outputs;
                                outputParamsTypes = _.map(outputAbi, 'type');
                                resultArray = ethersContracts.Interface.decodeParams(outputParamsTypes, rawCallResult);
                                resultArray = base_contract_1.BaseContract._transformABIData(outputAbi, resultArray, base_contract_1.BaseContract._lowercaseAddress.bind(this));
                                return [2 /*return*/, resultArray[0]];
                        }
                    });
                });
            },
        };
        _this.TOKEN_TRANSFER_PROXY_CONTRACT = {
            callAsync: function (txData, defaultBlock) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, callData, rawCallResult, outputAbi, outputParamsTypes, resultArray;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'TOKEN_TRANSFER_PROXY_CONTRACT' }).inputs;
                                base_contract_1.BaseContract._transformABIData(inputAbi, [], base_contract_1.BaseContract._bigNumberToString.bind(this));
                                encodedData = self._ethersInterface.functions.TOKEN_TRANSFER_PROXY_CONTRACT().data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync({
                                        data: encodedData,
                                    })];
                            case 1:
                                callData = _a.sent();
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callData, defaultBlock)];
                            case 2:
                                rawCallResult = _a.sent();
                                outputAbi = _.find(this.abi, { name: 'TOKEN_TRANSFER_PROXY_CONTRACT' }).outputs;
                                outputParamsTypes = _.map(outputAbi, 'type');
                                resultArray = ethersContracts.Interface.decodeParams(outputParamsTypes, rawCallResult);
                                resultArray = base_contract_1.BaseContract._transformABIData(outputAbi, resultArray, base_contract_1.BaseContract._lowercaseAddress.bind(this));
                                return [2 /*return*/, resultArray[0]];
                        }
                    });
                });
            },
        };
        _this.batchFillOrders = {
            sendTransactionAsync: function (orderAddresses, orderValues, fillTakerTokenAmounts, shouldThrowOnInsufficientBalanceOrAllowance, v, r, s, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, txDataWithDefaults, txHash, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'batchFillOrders' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [
                                    orderAddresses,
                                    orderValues,
                                    fillTakerTokenAmounts,
                                    shouldThrowOnInsufficientBalanceOrAllowance,
                                    v,
                                    r,
                                    s,
                                ], base_contract_1.BaseContract._bigNumberToString.bind(this)), orderAddresses = _a[0], orderValues = _a[1], fillTakerTokenAmounts = _a[2], shouldThrowOnInsufficientBalanceOrAllowance = _a[3], v = _a[4], r = _a[5], s = _a[6];
                                encodedData = this._ethersInterface.functions.batchFillOrders(orderAddresses, orderValues, fillTakerTokenAmounts, shouldThrowOnInsufficientBalanceOrAllowance, v, r, s).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync(__assign({}, txData, { data: encodedData }), self.batchFillOrders.estimateGasAsync.bind(self, orderAddresses, orderValues, fillTakerTokenAmounts, shouldThrowOnInsufficientBalanceOrAllowance, v, r, s))];
                            case 1:
                                txDataWithDefaults = _b.sent();
                                return [4 /*yield*/, this._web3Wrapper.sendTransactionAsync(txDataWithDefaults)];
                            case 2:
                                txHash = _b.sent();
                                return [2 /*return*/, txHash];
                        }
                    });
                });
            },
            estimateGasAsync: function (orderAddresses, orderValues, fillTakerTokenAmounts, shouldThrowOnInsufficientBalanceOrAllowance, v, r, s, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, txDataWithDefaults, gas, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'batchFillOrders' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [
                                    orderAddresses,
                                    orderValues,
                                    fillTakerTokenAmounts,
                                    shouldThrowOnInsufficientBalanceOrAllowance,
                                    v,
                                    r,
                                    s,
                                ], base_contract_1.BaseContract._bigNumberToString.bind(this)), orderAddresses = _a[0], orderValues = _a[1], fillTakerTokenAmounts = _a[2], shouldThrowOnInsufficientBalanceOrAllowance = _a[3], v = _a[4], r = _a[5], s = _a[6];
                                encodedData = this._ethersInterface.functions.batchFillOrders(orderAddresses, orderValues, fillTakerTokenAmounts, shouldThrowOnInsufficientBalanceOrAllowance, v, r, s).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync(__assign({}, txData, { data: encodedData }))];
                            case 1:
                                txDataWithDefaults = _b.sent();
                                return [4 /*yield*/, this._web3Wrapper.estimateGasAsync(txDataWithDefaults)];
                            case 2:
                                gas = _b.sent();
                                return [2 /*return*/, gas];
                        }
                    });
                });
            },
            getABIEncodedTransactionData: function (orderAddresses, orderValues, fillTakerTokenAmounts, shouldThrowOnInsufficientBalanceOrAllowance, v, r, s, txData) {
                if (txData === void 0) { txData = {}; }
                var self = this;
                var inputAbi = _.find(this.abi, { name: 'batchFillOrders' }).inputs;
                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [
                    orderAddresses,
                    orderValues,
                    fillTakerTokenAmounts,
                    shouldThrowOnInsufficientBalanceOrAllowance,
                    v,
                    r,
                    s,
                ], base_contract_1.BaseContract._bigNumberToString.bind(this)), orderAddresses = _a[0], orderValues = _a[1], fillTakerTokenAmounts = _a[2], shouldThrowOnInsufficientBalanceOrAllowance = _a[3], v = _a[4], r = _a[5], s = _a[6];
                var abiEncodedTransactionData = this._ethersInterface.functions.batchFillOrders(orderAddresses, orderValues, fillTakerTokenAmounts, shouldThrowOnInsufficientBalanceOrAllowance, v, r, s).data;
                return abiEncodedTransactionData;
                var _a;
            },
        };
        _this.batchCancelOrders = {
            sendTransactionAsync: function (orderAddresses, orderValues, cancelTakerTokenAmounts, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, txDataWithDefaults, txHash, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'batchCancelOrders' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [orderAddresses, orderValues, cancelTakerTokenAmounts], base_contract_1.BaseContract._bigNumberToString.bind(this)), orderAddresses = _a[0], orderValues = _a[1], cancelTakerTokenAmounts = _a[2];
                                encodedData = this._ethersInterface.functions.batchCancelOrders(orderAddresses, orderValues, cancelTakerTokenAmounts).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync(__assign({}, txData, { data: encodedData }), self.batchCancelOrders.estimateGasAsync.bind(self, orderAddresses, orderValues, cancelTakerTokenAmounts))];
                            case 1:
                                txDataWithDefaults = _b.sent();
                                return [4 /*yield*/, this._web3Wrapper.sendTransactionAsync(txDataWithDefaults)];
                            case 2:
                                txHash = _b.sent();
                                return [2 /*return*/, txHash];
                        }
                    });
                });
            },
            estimateGasAsync: function (orderAddresses, orderValues, cancelTakerTokenAmounts, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, txDataWithDefaults, gas, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'batchCancelOrders' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [orderAddresses, orderValues, cancelTakerTokenAmounts], base_contract_1.BaseContract._bigNumberToString.bind(this)), orderAddresses = _a[0], orderValues = _a[1], cancelTakerTokenAmounts = _a[2];
                                encodedData = this._ethersInterface.functions.batchCancelOrders(orderAddresses, orderValues, cancelTakerTokenAmounts).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync(__assign({}, txData, { data: encodedData }))];
                            case 1:
                                txDataWithDefaults = _b.sent();
                                return [4 /*yield*/, this._web3Wrapper.estimateGasAsync(txDataWithDefaults)];
                            case 2:
                                gas = _b.sent();
                                return [2 /*return*/, gas];
                        }
                    });
                });
            },
            getABIEncodedTransactionData: function (orderAddresses, orderValues, cancelTakerTokenAmounts, txData) {
                if (txData === void 0) { txData = {}; }
                var self = this;
                var inputAbi = _.find(this.abi, { name: 'batchCancelOrders' }).inputs;
                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [orderAddresses, orderValues, cancelTakerTokenAmounts], base_contract_1.BaseContract._bigNumberToString.bind(this)), orderAddresses = _a[0], orderValues = _a[1], cancelTakerTokenAmounts = _a[2];
                var abiEncodedTransactionData = this._ethersInterface.functions.batchCancelOrders(orderAddresses, orderValues, cancelTakerTokenAmounts).data;
                return abiEncodedTransactionData;
                var _a;
            },
        };
        _this.fillOrder = {
            sendTransactionAsync: function (orderAddresses, orderValues, fillTakerTokenAmount, shouldThrowOnInsufficientBalanceOrAllowance, v, r, s, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, txDataWithDefaults, txHash, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'fillOrder' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [
                                    orderAddresses,
                                    orderValues,
                                    fillTakerTokenAmount,
                                    shouldThrowOnInsufficientBalanceOrAllowance,
                                    v,
                                    r,
                                    s,
                                ], base_contract_1.BaseContract._bigNumberToString.bind(this)), orderAddresses = _a[0], orderValues = _a[1], fillTakerTokenAmount = _a[2], shouldThrowOnInsufficientBalanceOrAllowance = _a[3], v = _a[4], r = _a[5], s = _a[6];
                                encodedData = this._ethersInterface.functions.fillOrder(orderAddresses, orderValues, fillTakerTokenAmount, shouldThrowOnInsufficientBalanceOrAllowance, v, r, s).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync(__assign({}, txData, { data: encodedData }), self.fillOrder.estimateGasAsync.bind(self, orderAddresses, orderValues, fillTakerTokenAmount, shouldThrowOnInsufficientBalanceOrAllowance, v, r, s))];
                            case 1:
                                txDataWithDefaults = _b.sent();
                                return [4 /*yield*/, this._web3Wrapper.sendTransactionAsync(txDataWithDefaults)];
                            case 2:
                                txHash = _b.sent();
                                return [2 /*return*/, txHash];
                        }
                    });
                });
            },
            estimateGasAsync: function (orderAddresses, orderValues, fillTakerTokenAmount, shouldThrowOnInsufficientBalanceOrAllowance, v, r, s, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, txDataWithDefaults, gas, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'fillOrder' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [
                                    orderAddresses,
                                    orderValues,
                                    fillTakerTokenAmount,
                                    shouldThrowOnInsufficientBalanceOrAllowance,
                                    v,
                                    r,
                                    s,
                                ], base_contract_1.BaseContract._bigNumberToString.bind(this)), orderAddresses = _a[0], orderValues = _a[1], fillTakerTokenAmount = _a[2], shouldThrowOnInsufficientBalanceOrAllowance = _a[3], v = _a[4], r = _a[5], s = _a[6];
                                encodedData = this._ethersInterface.functions.fillOrder(orderAddresses, orderValues, fillTakerTokenAmount, shouldThrowOnInsufficientBalanceOrAllowance, v, r, s).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync(__assign({}, txData, { data: encodedData }))];
                            case 1:
                                txDataWithDefaults = _b.sent();
                                return [4 /*yield*/, this._web3Wrapper.estimateGasAsync(txDataWithDefaults)];
                            case 2:
                                gas = _b.sent();
                                return [2 /*return*/, gas];
                        }
                    });
                });
            },
            getABIEncodedTransactionData: function (orderAddresses, orderValues, fillTakerTokenAmount, shouldThrowOnInsufficientBalanceOrAllowance, v, r, s, txData) {
                if (txData === void 0) { txData = {}; }
                var self = this;
                var inputAbi = _.find(this.abi, { name: 'fillOrder' }).inputs;
                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [
                    orderAddresses,
                    orderValues,
                    fillTakerTokenAmount,
                    shouldThrowOnInsufficientBalanceOrAllowance,
                    v,
                    r,
                    s,
                ], base_contract_1.BaseContract._bigNumberToString.bind(this)), orderAddresses = _a[0], orderValues = _a[1], fillTakerTokenAmount = _a[2], shouldThrowOnInsufficientBalanceOrAllowance = _a[3], v = _a[4], r = _a[5], s = _a[6];
                var abiEncodedTransactionData = this._ethersInterface.functions.fillOrder(orderAddresses, orderValues, fillTakerTokenAmount, shouldThrowOnInsufficientBalanceOrAllowance, v, r, s).data;
                return abiEncodedTransactionData;
                var _a;
            },
            callAsync: function (orderAddresses, orderValues, fillTakerTokenAmount, shouldThrowOnInsufficientBalanceOrAllowance, v, r, s, txData, defaultBlock) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, callData, rawCallResult, outputAbi, outputParamsTypes, resultArray, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'fillOrder' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [
                                    orderAddresses,
                                    orderValues,
                                    fillTakerTokenAmount,
                                    shouldThrowOnInsufficientBalanceOrAllowance,
                                    v,
                                    r,
                                    s,
                                ], base_contract_1.BaseContract._bigNumberToString.bind(this)), orderAddresses = _a[0], orderValues = _a[1], fillTakerTokenAmount = _a[2], shouldThrowOnInsufficientBalanceOrAllowance = _a[3], v = _a[4], r = _a[5], s = _a[6];
                                encodedData = self._ethersInterface.functions.fillOrder(orderAddresses, orderValues, fillTakerTokenAmount, shouldThrowOnInsufficientBalanceOrAllowance, v, r, s).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync({
                                        data: encodedData,
                                    })];
                            case 1:
                                callData = _b.sent();
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callData, defaultBlock)];
                            case 2:
                                rawCallResult = _b.sent();
                                outputAbi = _.find(this.abi, { name: 'fillOrder' }).outputs;
                                outputParamsTypes = _.map(outputAbi, 'type');
                                resultArray = ethersContracts.Interface.decodeParams(outputParamsTypes, rawCallResult);
                                resultArray = base_contract_1.BaseContract._transformABIData(outputAbi, resultArray, base_contract_1.BaseContract._lowercaseAddress.bind(this));
                                return [2 /*return*/, resultArray[0]];
                        }
                    });
                });
            },
        };
        _this.getOrderHash = {
            callAsync: function (orderAddresses, orderValues, txData, defaultBlock) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, callData, rawCallResult, outputAbi, outputParamsTypes, resultArray, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'getOrderHash' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [orderAddresses, orderValues], base_contract_1.BaseContract._bigNumberToString.bind(this)), orderAddresses = _a[0], orderValues = _a[1];
                                encodedData = self._ethersInterface.functions.getOrderHash(orderAddresses, orderValues).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync({
                                        data: encodedData,
                                    })];
                            case 1:
                                callData = _b.sent();
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callData, defaultBlock)];
                            case 2:
                                rawCallResult = _b.sent();
                                outputAbi = _.find(this.abi, { name: 'getOrderHash' }).outputs;
                                outputParamsTypes = _.map(outputAbi, 'type');
                                resultArray = ethersContracts.Interface.decodeParams(outputParamsTypes, rawCallResult);
                                resultArray = base_contract_1.BaseContract._transformABIData(outputAbi, resultArray, base_contract_1.BaseContract._lowercaseAddress.bind(this));
                                return [2 /*return*/, resultArray[0]];
                        }
                    });
                });
            },
        };
        _this.EXTERNAL_QUERY_GAS_LIMIT = {
            callAsync: function (txData, defaultBlock) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, callData, rawCallResult, outputAbi, outputParamsTypes, resultArray;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'EXTERNAL_QUERY_GAS_LIMIT' }).inputs;
                                base_contract_1.BaseContract._transformABIData(inputAbi, [], base_contract_1.BaseContract._bigNumberToString.bind(this));
                                encodedData = self._ethersInterface.functions.EXTERNAL_QUERY_GAS_LIMIT().data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync({
                                        data: encodedData,
                                    })];
                            case 1:
                                callData = _a.sent();
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callData, defaultBlock)];
                            case 2:
                                rawCallResult = _a.sent();
                                outputAbi = _.find(this.abi, { name: 'EXTERNAL_QUERY_GAS_LIMIT' }).outputs;
                                outputParamsTypes = _.map(outputAbi, 'type');
                                resultArray = ethersContracts.Interface.decodeParams(outputParamsTypes, rawCallResult);
                                resultArray = base_contract_1.BaseContract._transformABIData(outputAbi, resultArray, base_contract_1.BaseContract._lowercaseAddress.bind(this));
                                return [2 /*return*/, resultArray[0]];
                        }
                    });
                });
            },
        };
        _this.VERSION = {
            callAsync: function (txData, defaultBlock) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, callData, rawCallResult, outputAbi, outputParamsTypes, resultArray;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'VERSION' }).inputs;
                                base_contract_1.BaseContract._transformABIData(inputAbi, [], base_contract_1.BaseContract._bigNumberToString.bind(this));
                                encodedData = self._ethersInterface.functions.VERSION().data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync({
                                        data: encodedData,
                                    })];
                            case 1:
                                callData = _a.sent();
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callData, defaultBlock)];
                            case 2:
                                rawCallResult = _a.sent();
                                outputAbi = _.find(this.abi, { name: 'VERSION' }).outputs;
                                outputParamsTypes = _.map(outputAbi, 'type');
                                resultArray = ethersContracts.Interface.decodeParams(outputParamsTypes, rawCallResult);
                                resultArray = base_contract_1.BaseContract._transformABIData(outputAbi, resultArray, base_contract_1.BaseContract._lowercaseAddress.bind(this));
                                return [2 /*return*/, resultArray[0]];
                        }
                    });
                });
            },
        };
        utils_1.classUtils.bindAll(_this, ['_ethersInterface', 'address', 'abi', '_web3Wrapper']);
        return _this;
    }
    return ExchangeContract;
}(base_contract_1.BaseContract)); // tslint:disable:max-file-line-count
exports.ExchangeContract = ExchangeContract;
