// 0x.js/lib/src/contract_wrappers/generated/token.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * This file is auto-generated using abi-gen. Don't edit directly.
 * Templates can be found at https://github.com/0xProject/0x-monorepo/tree/development/packages/contract_templates.
 */
// tslint:disable:no-consecutive-blank-lines
// tslint:disable-next-line:no-unused-variable
var base_contract_1 = require("@0xproject/base-contract");
var utils_1 = require("@0xproject/utils");
var ethersContracts = require("ethers-contracts");
var _ = require("lodash");
var TokenEvents;
(function (TokenEvents) {
    TokenEvents["Transfer"] = "Transfer";
    TokenEvents["Approval"] = "Approval";
})(TokenEvents = exports.TokenEvents || (exports.TokenEvents = {}));
// tslint:disable:no-parameter-reassignment
var TokenContract = /** @class */ (function (_super) {
    __extends(TokenContract, _super);
    function TokenContract(web3Wrapper, abi, address) {
        var _this = _super.call(this, web3Wrapper, abi, address) || this;
        _this.approve = {
            sendTransactionAsync: function (_spender, _value, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, txDataWithDefaults, txHash, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'approve' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [_spender, _value], base_contract_1.BaseContract._bigNumberToString.bind(this)), _spender = _a[0], _value = _a[1];
                                encodedData = this._ethersInterface.functions.approve(_spender, _value).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync(__assign({}, txData, { data: encodedData }), self.approve.estimateGasAsync.bind(self, _spender, _value))];
                            case 1:
                                txDataWithDefaults = _b.sent();
                                return [4 /*yield*/, this._web3Wrapper.sendTransactionAsync(txDataWithDefaults)];
                            case 2:
                                txHash = _b.sent();
                                return [2 /*return*/, txHash];
                        }
                    });
                });
            },
            estimateGasAsync: function (_spender, _value, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, txDataWithDefaults, gas, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'approve' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [_spender, _value], base_contract_1.BaseContract._bigNumberToString.bind(this)), _spender = _a[0], _value = _a[1];
                                encodedData = this._ethersInterface.functions.approve(_spender, _value).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync(__assign({}, txData, { data: encodedData }))];
                            case 1:
                                txDataWithDefaults = _b.sent();
                                return [4 /*yield*/, this._web3Wrapper.estimateGasAsync(txDataWithDefaults)];
                            case 2:
                                gas = _b.sent();
                                return [2 /*return*/, gas];
                        }
                    });
                });
            },
            getABIEncodedTransactionData: function (_spender, _value, txData) {
                if (txData === void 0) { txData = {}; }
                var self = this;
                var inputAbi = _.find(this.abi, { name: 'approve' }).inputs;
                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [_spender, _value], base_contract_1.BaseContract._bigNumberToString.bind(this)), _spender = _a[0], _value = _a[1];
                var abiEncodedTransactionData = this._ethersInterface.functions.approve(_spender, _value).data;
                return abiEncodedTransactionData;
                var _a;
            },
            callAsync: function (_spender, _value, txData, defaultBlock) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, callData, rawCallResult, outputAbi, outputParamsTypes, resultArray, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'approve' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [_spender, _value], base_contract_1.BaseContract._bigNumberToString.bind(this)), _spender = _a[0], _value = _a[1];
                                encodedData = self._ethersInterface.functions.approve(_spender, _value).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync({
                                        data: encodedData,
                                    })];
                            case 1:
                                callData = _b.sent();
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callData, defaultBlock)];
                            case 2:
                                rawCallResult = _b.sent();
                                outputAbi = _.find(this.abi, { name: 'approve' }).outputs;
                                outputParamsTypes = _.map(outputAbi, 'type');
                                resultArray = ethersContracts.Interface.decodeParams(outputParamsTypes, rawCallResult);
                                resultArray = base_contract_1.BaseContract._transformABIData(outputAbi, resultArray, base_contract_1.BaseContract._lowercaseAddress.bind(this));
                                return [2 /*return*/, resultArray[0]];
                        }
                    });
                });
            },
        };
        _this.totalSupply = {
            callAsync: function (txData, defaultBlock) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, callData, rawCallResult, outputAbi, outputParamsTypes, resultArray;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'totalSupply' }).inputs;
                                base_contract_1.BaseContract._transformABIData(inputAbi, [], base_contract_1.BaseContract._bigNumberToString.bind(this));
                                encodedData = self._ethersInterface.functions.totalSupply().data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync({
                                        data: encodedData,
                                    })];
                            case 1:
                                callData = _a.sent();
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callData, defaultBlock)];
                            case 2:
                                rawCallResult = _a.sent();
                                outputAbi = _.find(this.abi, { name: 'totalSupply' }).outputs;
                                outputParamsTypes = _.map(outputAbi, 'type');
                                resultArray = ethersContracts.Interface.decodeParams(outputParamsTypes, rawCallResult);
                                resultArray = base_contract_1.BaseContract._transformABIData(outputAbi, resultArray, base_contract_1.BaseContract._lowercaseAddress.bind(this));
                                return [2 /*return*/, resultArray[0]];
                        }
                    });
                });
            },
        };
        _this.transferFrom = {
            sendTransactionAsync: function (_from, _to, _value, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, txDataWithDefaults, txHash, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'transferFrom' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [_from, _to, _value], base_contract_1.BaseContract._bigNumberToString.bind(this)), _from = _a[0], _to = _a[1], _value = _a[2];
                                encodedData = this._ethersInterface.functions.transferFrom(_from, _to, _value).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync(__assign({}, txData, { data: encodedData }), self.transferFrom.estimateGasAsync.bind(self, _from, _to, _value))];
                            case 1:
                                txDataWithDefaults = _b.sent();
                                return [4 /*yield*/, this._web3Wrapper.sendTransactionAsync(txDataWithDefaults)];
                            case 2:
                                txHash = _b.sent();
                                return [2 /*return*/, txHash];
                        }
                    });
                });
            },
            estimateGasAsync: function (_from, _to, _value, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, txDataWithDefaults, gas, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'transferFrom' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [_from, _to, _value], base_contract_1.BaseContract._bigNumberToString.bind(this)), _from = _a[0], _to = _a[1], _value = _a[2];
                                encodedData = this._ethersInterface.functions.transferFrom(_from, _to, _value).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync(__assign({}, txData, { data: encodedData }))];
                            case 1:
                                txDataWithDefaults = _b.sent();
                                return [4 /*yield*/, this._web3Wrapper.estimateGasAsync(txDataWithDefaults)];
                            case 2:
                                gas = _b.sent();
                                return [2 /*return*/, gas];
                        }
                    });
                });
            },
            getABIEncodedTransactionData: function (_from, _to, _value, txData) {
                if (txData === void 0) { txData = {}; }
                var self = this;
                var inputAbi = _.find(this.abi, { name: 'transferFrom' }).inputs;
                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [_from, _to, _value], base_contract_1.BaseContract._bigNumberToString.bind(this)), _from = _a[0], _to = _a[1], _value = _a[2];
                var abiEncodedTransactionData = this._ethersInterface.functions.transferFrom(_from, _to, _value).data;
                return abiEncodedTransactionData;
                var _a;
            },
            callAsync: function (_from, _to, _value, txData, defaultBlock) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, callData, rawCallResult, outputAbi, outputParamsTypes, resultArray, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'transferFrom' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [_from, _to, _value], base_contract_1.BaseContract._bigNumberToString.bind(this)), _from = _a[0], _to = _a[1], _value = _a[2];
                                encodedData = self._ethersInterface.functions.transferFrom(_from, _to, _value).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync({
                                        data: encodedData,
                                    })];
                            case 1:
                                callData = _b.sent();
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callData, defaultBlock)];
                            case 2:
                                rawCallResult = _b.sent();
                                outputAbi = _.find(this.abi, { name: 'transferFrom' }).outputs;
                                outputParamsTypes = _.map(outputAbi, 'type');
                                resultArray = ethersContracts.Interface.decodeParams(outputParamsTypes, rawCallResult);
                                resultArray = base_contract_1.BaseContract._transformABIData(outputAbi, resultArray, base_contract_1.BaseContract._lowercaseAddress.bind(this));
                                return [2 /*return*/, resultArray[0]];
                        }
                    });
                });
            },
        };
        _this.balanceOf = {
            callAsync: function (_owner, txData, defaultBlock) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, callData, rawCallResult, outputAbi, outputParamsTypes, resultArray;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'balanceOf' }).inputs;
                                _owner = base_contract_1.BaseContract._transformABIData(inputAbi, [_owner], base_contract_1.BaseContract._bigNumberToString.bind(this))[0];
                                encodedData = self._ethersInterface.functions.balanceOf(_owner).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync({
                                        data: encodedData,
                                    })];
                            case 1:
                                callData = _a.sent();
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callData, defaultBlock)];
                            case 2:
                                rawCallResult = _a.sent();
                                outputAbi = _.find(this.abi, { name: 'balanceOf' }).outputs;
                                outputParamsTypes = _.map(outputAbi, 'type');
                                resultArray = ethersContracts.Interface.decodeParams(outputParamsTypes, rawCallResult);
                                resultArray = base_contract_1.BaseContract._transformABIData(outputAbi, resultArray, base_contract_1.BaseContract._lowercaseAddress.bind(this));
                                return [2 /*return*/, resultArray[0]];
                        }
                    });
                });
            },
        };
        _this.transfer = {
            sendTransactionAsync: function (_to, _value, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, txDataWithDefaults, txHash, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'transfer' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [_to, _value], base_contract_1.BaseContract._bigNumberToString.bind(this)), _to = _a[0], _value = _a[1];
                                encodedData = this._ethersInterface.functions.transfer(_to, _value).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync(__assign({}, txData, { data: encodedData }), self.transfer.estimateGasAsync.bind(self, _to, _value))];
                            case 1:
                                txDataWithDefaults = _b.sent();
                                return [4 /*yield*/, this._web3Wrapper.sendTransactionAsync(txDataWithDefaults)];
                            case 2:
                                txHash = _b.sent();
                                return [2 /*return*/, txHash];
                        }
                    });
                });
            },
            estimateGasAsync: function (_to, _value, txData) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, txDataWithDefaults, gas, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'transfer' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [_to, _value], base_contract_1.BaseContract._bigNumberToString.bind(this)), _to = _a[0], _value = _a[1];
                                encodedData = this._ethersInterface.functions.transfer(_to, _value).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync(__assign({}, txData, { data: encodedData }))];
                            case 1:
                                txDataWithDefaults = _b.sent();
                                return [4 /*yield*/, this._web3Wrapper.estimateGasAsync(txDataWithDefaults)];
                            case 2:
                                gas = _b.sent();
                                return [2 /*return*/, gas];
                        }
                    });
                });
            },
            getABIEncodedTransactionData: function (_to, _value, txData) {
                if (txData === void 0) { txData = {}; }
                var self = this;
                var inputAbi = _.find(this.abi, { name: 'transfer' }).inputs;
                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [_to, _value], base_contract_1.BaseContract._bigNumberToString.bind(this)), _to = _a[0], _value = _a[1];
                var abiEncodedTransactionData = this._ethersInterface.functions.transfer(_to, _value).data;
                return abiEncodedTransactionData;
                var _a;
            },
            callAsync: function (_to, _value, txData, defaultBlock) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, callData, rawCallResult, outputAbi, outputParamsTypes, resultArray, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'transfer' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [_to, _value], base_contract_1.BaseContract._bigNumberToString.bind(this)), _to = _a[0], _value = _a[1];
                                encodedData = self._ethersInterface.functions.transfer(_to, _value).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync({
                                        data: encodedData,
                                    })];
                            case 1:
                                callData = _b.sent();
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callData, defaultBlock)];
                            case 2:
                                rawCallResult = _b.sent();
                                outputAbi = _.find(this.abi, { name: 'transfer' }).outputs;
                                outputParamsTypes = _.map(outputAbi, 'type');
                                resultArray = ethersContracts.Interface.decodeParams(outputParamsTypes, rawCallResult);
                                resultArray = base_contract_1.BaseContract._transformABIData(outputAbi, resultArray, base_contract_1.BaseContract._lowercaseAddress.bind(this));
                                return [2 /*return*/, resultArray[0]];
                        }
                    });
                });
            },
        };
        _this.allowance = {
            callAsync: function (_owner, _spender, txData, defaultBlock) {
                if (txData === void 0) { txData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, inputAbi, encodedData, callData, rawCallResult, outputAbi, outputParamsTypes, resultArray, _a;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                self = this;
                                inputAbi = _.find(this.abi, { name: 'allowance' }).inputs;
                                _a = base_contract_1.BaseContract._transformABIData(inputAbi, [_owner, _spender], base_contract_1.BaseContract._bigNumberToString.bind(this)), _owner = _a[0], _spender = _a[1];
                                encodedData = self._ethersInterface.functions.allowance(_owner, _spender).data;
                                return [4 /*yield*/, self._applyDefaultsToTxDataAsync({
                                        data: encodedData,
                                    })];
                            case 1:
                                callData = _b.sent();
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callData, defaultBlock)];
                            case 2:
                                rawCallResult = _b.sent();
                                outputAbi = _.find(this.abi, { name: 'allowance' }).outputs;
                                outputParamsTypes = _.map(outputAbi, 'type');
                                resultArray = ethersContracts.Interface.decodeParams(outputParamsTypes, rawCallResult);
                                resultArray = base_contract_1.BaseContract._transformABIData(outputAbi, resultArray, base_contract_1.BaseContract._lowercaseAddress.bind(this));
                                return [2 /*return*/, resultArray[0]];
                        }
                    });
                });
            },
        };
        utils_1.classUtils.bindAll(_this, ['_ethersInterface', 'address', 'abi', '_web3Wrapper']);
        return _this;
    }
    return TokenContract;
}(base_contract_1.BaseContract)); // tslint:disable:max-file-line-count
exports.TokenContract = TokenContract;
