// 0x.js/lib/src/schemas/zero_ex_private_network_config_schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.zeroExPrivateNetworkConfigSchema = {
    id: '/ZeroExPrivateNetworkConfig',
    properties: {
        networkId: {
            type: 'number',
            minimum: 1,
        },
        gasPrice: { $ref: '/Number' },
        zrxContractAddress: { $ref: '/Address' },
        exchangeContractAddress: { $ref: '/Address' },
        tokenRegistryContractAddress: { $ref: '/Address' },
        tokenTransferProxyContractAddress: { $ref: '/Address' },
        orderWatcherConfig: {
            type: 'object',
            properties: {
                pollingIntervalMs: {
                    type: 'number',
                    minimum: 0,
                },
                numConfirmations: {
                    type: 'number',
                    minimum: 0,
                },
            },
        },
    },
    type: 'object',
    required: [
        'networkId',
        'zrxContractAddress',
        'exchangeContractAddress',
        'tokenRegistryContractAddress',
        'tokenTransferProxyContractAddress',
    ],
};
