// ddex-api/dist/models/order.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _0x_js_1 = require("0x.js");
var types_1 = require("../types");
var bignumber_js_1 = require("bignumber.js");
var Order = /** @class */ (function () {
    function Order() {
    }
    Order.fromJSON = function (orderJson) {
        var order = new Order();
        order.id = orderJson.id;
        order.pair = orderJson.pair;
        order.status = orderJson.status;
        order.account = orderJson.account;
        if (orderJson.side === "sell") {
            order.side = types_1.Side.SELL;
        }
        else if (orderJson.side === "buy") {
            order.side = types_1.Side.BUY;
        }
        order.amount = new bignumber_js_1.BigNumber(orderJson.amount);
        order.price = new bignumber_js_1.BigNumber(orderJson.price);
        order.fee = new bignumber_js_1.BigNumber(orderJson.fee);
        order.availableAmount = new bignumber_js_1.BigNumber(orderJson.availableAmount);
        order.pendingAmount = new bignumber_js_1.BigNumber(orderJson.pendingAmount);
        order.canceledAmount = new bignumber_js_1.BigNumber(orderJson.canceledAmount);
        order.confirmedAmount = new bignumber_js_1.BigNumber(orderJson.confirmedAmount);
        order.createdAt = new Date(orderJson.createdAt);
        order.json = {
            maker: orderJson.json.maker,
            taker: orderJson.json.taker,
            makerFee: new bignumber_js_1.BigNumber(orderJson.json.makerFee),
            takerFee: new bignumber_js_1.BigNumber(orderJson.json.takerFee),
            makerTokenAmount: new bignumber_js_1.BigNumber(orderJson.json.makerTokenAmount),
            takerTokenAmount: new bignumber_js_1.BigNumber(orderJson.json.takerTokenAmount),
            makerTokenAddress: orderJson.json.makerTokenAddress,
            takerTokenAddress: orderJson.json.takerTokenAddress,
            salt: _0x_js_1.ZeroEx.generatePseudoRandomSalt(),
            exchangeContractAddress: orderJson.json.exchangeContractAddress,
            feeRecipient: orderJson.json.feeRecipient,
            expirationUnixTimestampSec: new bignumber_js_1.BigNumber(orderJson.json.expirationUnixTimestampSec)
        };
        return order;
    };
    return Order;
}());
exports.Order = Order;
